docs/Code-Base-Walkthrough.md

```diff
@@ -0,0 +1,200 @@
+
+**Testing**
+
+- [Jest ](https://jestjs.io/docs/en/getting-started) Unit tests are run with `npm run test`
+- [Spectron](https://www.electronjs.org/spectron) Used for integration testing `npm run itest`
+- [Enzyme](https://enzymejs.github.io/enzyme/docs/api/) Used only a little for testing Components
+
+
+
+## Patterns
+
+**Closures over Classes**
+
+For utility classes and domain models, we've been using functions that return objects with methods and data, rather than the `Class`. Prefer this style over classes if possible. 
+
+```js
+// Closure for example (preferred)
+function createLog(data) {
+  return {
+    getField: (name) => data.find(f => f.name === name)
+  }
+}
+```
+
+```js
+// Class for example (not preferred)
+Class Log {
+  constructor(data) {
+    this.data = data
+  }
+  
+  getFieid(name) {
+    return this.data.find(f => f.name === name)
+  }
+}
+```
+
+
+
+**Testing**
+
+Tests go right next to file they are testing with a `.test.js` suffix. Tests can accompany:
+
+* State slices (actions, selectors)
+* Business logic
+* Domain classes
+* Utility classes
+* React Components
+
+However, it's not very easy to test React components in unit tests, so we don't have many. The integration tests help with coverage there.
+
+**Hooks**
+
+There are some useful, generic hooks located in `src/js/components/hooks`.
+
+
+## Diverging Patterns
+
+Some areas of the code do the same thing in two different ways. It can be difficult for new developers to know which pattern to choose. Hopefully this will help.
+
+**Models**
+
+Don't put new classes in the `src/js/models` directory. Instead, put domain classes in the `src/js/brim` directory and put utility models in the `src/js/lib/` directory. The models directory used to contain all domain and utility classes. Note that the structures in `src/js/lib` are also in transition. There are some domain specific ones there that should be moved.
+
+**Styles**
+
+Use the Styled Components library to style new components. Previously, we used scss files located in `src/css`. Many of the components are styled with scss, and class names, but we recently committed to Styled Components. We also have a "theme" that holds all the common colors and styles used in our UI.
+
+**Functional Components**
+
+Do not make new components using the React "Class" component API. Instead use the functional component API.
+
+**Animation**
+
+There is no solid convention on how to create animations. We've attempted `anime.js`, `react-spring.js`, native Web Animations , and css animation/transition properties. New code can use any method they prefer and seems appropriate. I'd say avoid react-spring because there are lots of API changes coming in v9. In the future, for drag and drop (like dragging around the tabs), I'd like to try out `react-beautiful-dnd`. I think that could replace our usage of `react-spring` and make the code much more readable.
```

Comment:
```suggestion
There is no solid convention on how to create animations. We've attempted `anime.js`, `react-spring.js`, native Web Animations, and css animation/transition properties. New code can use any method they prefer and seems appropriate. Avoid react-spring because there are lots of API changes coming in v9. In the future, for drag and drop (like dragging around the tabs), we'd like to try out `react-beautiful-dnd`, which could replace our usage of `react-spring` and make the code more readable.
```